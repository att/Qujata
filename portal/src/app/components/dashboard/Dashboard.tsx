/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/typedef */
/* eslint-disable jsx-a11y/label-has-associated-control */
import { max, mean, min, round } from 'lodash';
import React, { NamedExoticComponent } from 'react';
import { ChartDataMap, ITestResponseData, ITimeData } from '../../shared/models/quantum.interface';
import { Charts } from './components/charts';
import { MainBoxes } from './components/main-boxes';

import styles from './Dashboard.module.scss';

export interface DashboardProps {
  data: ChartDataMap;
}

function getTimeData(samples: ITestResponseData): ITimeData {
  if (!samples.totalTime && !samples.connectTime && !samples.downloadSpeed) {
    return {
      totalTime: { min: '0', avg: '0', max: '0' },
      connectTime: { min: '0', avg: '0', max: '0' },
      downloadSpeed: { min: '0', avg: '0', max: '0' },
    };
  }
  const { totalTime: totals, connectTime: connectTimes, downloadSpeed: downloadSpeeds } = samples;

  return {
    totalTime: {
      min: (Number((min(totals) as number) * 1000).toFixed(2)),
      avg: round(mean(totals) * 1000, 2).toString(),
      max: (Number((max(totals) as number) * 1000).toFixed(2)),
    },
    connectTime: {
      min: (Number((min(connectTimes) as number) * 1000).toFixed(2)),
      avg: round(mean(connectTimes) * 1000, 2).toString(),
      max: (Number((max(connectTimes) as number) * 1000).toFixed(2)),
    },
    downloadSpeed: {
      min: (min(downloadSpeeds) as number / 1024).toFixed(2),
      avg: (mean(downloadSpeeds) as number / 1024).toFixed(2),
      max: (max(downloadSpeeds) as number / 1024).toFixed(2),
    },
  };
}
export const Dashboard: React.FC<DashboardProps> = (props: DashboardProps) => {
  const { data } = props;
  const chartDataKeys: string[] = [];
  const chartData: Map<string, ITestResponseData | undefined> = new Map<string, ITestResponseData | undefined>();
  data.forEach((value, key) => {
    chartDataKeys.push(key.label);
    chartData.set(key.label, value);
  });

  return (
    <div className={styles.wrapper}>
      {chartDataKeys.map((key: string) => renderAlgorithmMainBoxes(key, chartData))}
      <Charts data={data} />
    </div>
  );
};

function renderAlgorithmMainBoxes(key:string, chartData: Map<string, ITestResponseData | undefined>) {
  const timeData: ITimeData = getTimeData(chartData.get(key) as ITestResponseData);
  return (
    <MainBoxes title={key} totalTime={timeData.totalTime} connectTime={timeData.connectTime} downloadSpeed={timeData.downloadSpeed} />
  );
}

export const DashboardMemoized: NamedExoticComponent<DashboardProps> = React.memo<DashboardProps>(Dashboard);
