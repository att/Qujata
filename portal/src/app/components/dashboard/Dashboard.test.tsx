import '@testing-library/jest-dom';
import { RenderResult, render, screen } from '@testing-library/react';
import { Dashboard, DashboardProps } from './Dashboard';
import { ITestParams } from '../../shared/models/quantum.interface';

// Mock the Charts component
jest.mock('./components/charts/Charts', () => ({
  Charts: () => <div data-testid="mocked-charts" />,
}));

describe('Dashboard', () => {
  let props: DashboardProps;
  let sampleTestParams: ITestParams;
  beforeAll(() => {
    // Prepare the props for the Dashboard component
    sampleTestParams = {
      experimentName: 'test',
      algorithms: {
        label: 'Sample label',
        value: 'Sample value',
      },
      iterationsCount: {
        label: '1000',
        value: '1000',
      },
      messageSizes: {
        label: '1024',
        value: '1024'
      },
      description: 'Sample description',
    };

    const sampleData = {
      totalTime: [0.1, 0.2, 0.3],
      connectTime: [0.4, 0.5, 0.6],
      downloadSpeed: [1024, 2048, 3072],
    };

    const data = new Map();
    data.set(sampleTestParams, sampleData);

    props = { data };
  });

  test('should render Dashboard', () => {
    // Render the Dashboard component with the prepared props
    const { container }: RenderResult = render(<Dashboard {...props} />);

    // Snapshot Matching
    expect(container.firstChild).toMatchSnapshot();
  });


  test('should render Dashboard, even when values are empty', () => {
    const data = new Map();
    data.set(sampleTestParams, {});

    props = { data };

    // Render the Dashboard component with the prepared props
    const { container }: RenderResult = render(<Dashboard {...props} />);

    // Snapshot Matching
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should check all values were rendered', () => {
    // Render the Dashboard component with the prepared props
    render(<Dashboard {...props} />);

    // Assert that the mocked MainBoxes component is rendered with the correct titles
    expect(screen.getByText('Session Time - Avg (mSec)')).toBeInTheDocument();
    expect(screen.getByText('Session Handshake Time - Avg (mSec)')).toBeInTheDocument();
    expect(screen.getByText('Download Speed - Avg (kB/sec)')).toBeInTheDocument();

    // Assert that the mocked Charts component is rendered
    expect(screen.getByTestId('mocked-charts')).toBeInTheDocument();
  });
});
