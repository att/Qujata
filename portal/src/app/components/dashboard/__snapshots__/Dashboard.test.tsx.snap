// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render Dashboard 1`] = `
<div
  class="wrapper"
>
  <div>
    <h3 />
    <div
      class="wrapper"
    >
      <div
        class="main_box_wrapper"
      >
        <div
          class="main_box_title"
        >
          Session Time - Avg (mSec)
        </div>
        <div
          class="main_box_value"
        >
          200
        </div>
        <div
          class="main_box_minmax"
        >
           
          <span>
             min: 
            100.00
              max: 
            300.00
          </span>
           
        </div>
      </div>
      <div
        class="main_box_wrapper"
      >
        <div
          class="main_box_title"
        >
          Session Handshake Time - Avg (mSec)
        </div>
        <div
          class="main_box_value"
        >
          500
        </div>
        <div
          class="main_box_minmax"
        >
           
          <span>
             min: 
            400.00
              max: 
            600.00
          </span>
           
        </div>
      </div>
      <div
        class="main_box_wrapper"
      >
        <div
          class="main_box_title"
        >
          Download Speed - Avg (kB/sec)
        </div>
        <div
          class="main_box_value"
        >
          2.00
        </div>
        <div
          class="main_box_minmax"
        >
           
          <span>
             min: 
            1.00
              max: 
            3.00
          </span>
           
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="mocked-charts"
  />
</div>
`;

exports[`Dashboard should render Dashboard, even when values are empty 1`] = `
<div
  class="wrapper"
>
  <div>
    <h3 />
    <div
      class="wrapper"
    >
      <div
        class="main_box_wrapper"
      >
        <div
          class="main_box_title"
        >
          Session Time - Avg (mSec)
        </div>
        <div
          class="main_box_value"
        >
          0
        </div>
        <div
          class="main_box_minmax"
        >
           
          <span>
             min: 
            0
              max: 
            0
          </span>
           
        </div>
      </div>
      <div
        class="main_box_wrapper"
      >
        <div
          class="main_box_title"
        >
          Session Handshake Time - Avg (mSec)
        </div>
        <div
          class="main_box_value"
        >
          0
        </div>
        <div
          class="main_box_minmax"
        >
           
          <span>
             min: 
            0
              max: 
            0
          </span>
           
        </div>
      </div>
      <div
        class="main_box_wrapper"
      >
        <div
          class="main_box_title"
        >
          Download Speed - Avg (kB/sec)
        </div>
        <div
          class="main_box_value"
        >
          0
        </div>
        <div
          class="main_box_minmax"
        >
           
          <span>
             min: 
            0
              max: 
            0
          </span>
           
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="mocked-charts"
  />
</div>
`;
